src/Changelog/UseKeepAChangelogEventsToReleaseAndFetchChangelog.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Laminas\AutomaticReleases\Changelog;
+
+use Laminas\AutomaticReleases\Git\CommitFile;
+use Laminas\AutomaticReleases\Git\Push;
+use Phly\KeepAChangelog\Version\ReadyLatestChangelogEvent;
+use Psr\EventDispatcher\EventDispatcherInterface;
+
+use function assert;
+use function date;
+use function file_exists;
+use function sprintf;
+
+class UseKeepAChangelogEventsToReleaseAndFetchChangelog implements
```

Comment:
The `CreateChangelogViaMilestone` and `UseKeepAChangelogEventsToReleaseAndFetchChangelog` don't really do the same things:

 * one creates a commit on the repo (this class)
 * the other (`UseKeepAChangelogEventsToReleaseAndFetchChangelog`) produces text to be used for the release

They should not share the same interface, and should not be put together under `ReleaseChangelogAndFetchContentsAggregate`.

In fact, the nullability in `    public function __invoke(ReleaseChangelogEvent $releaseChangelogEvent): ?string` indicates profound differences in behavior, and force you to implement a fallback mechanism if generating a changelog failed.

I think we should:

 1. use keep a changelog to generate `CHANGELOG.md` (and we do use it explicitly before creating a tag)
 2. use `CreateChangelogViaMilestone` to generate the tag description and release notes (same as it was before, relatively UI friendly for github, and still references a commit that contains a `CHANGELOG.md` diff)

Alternatively, we can either:

 * append `CreateChangelogViaMilestone` to tag description/release name, in addition to what `UseKeepAChangelogEventsToReleaseAndFetchChangelog` generated
 * scrap `CreateChangelogViaMilestone`, and generate everything with `UseKeepAChangelogEventsToReleaseAndFetchChangelog`

This means reverting some of these abstractions, and calling the logic sequentially:

```php
$keepAChangelogText = ($this->keepAChangelog)(...);
$releaseText = ($this->createChangelogViaMilestone)(...);

$tag(..., $releaseText . "\n\n" . $keepAChangelogText);
```

No need to aggregate these disjunct pieces of logic together, and no need to change pre-existing tests (besides CLI command tests)

Reply:
I disagree with some of this assessment.

KeepAChangelog does not **generate** a changelog; it **is** the changelog. The phly/keep-a-changelog tooling was being used to:

- Set the release date for the given changelog version.
- Retrieve the contents for that changelog version for purposes of the GitHub release.

That first part is where we have an overlap of responsibilities, as it also involves creating a commit.

The reason I want the Keep-A-Changelog format in the GitHub release is because if you search for `{package} {version} changelog`, you end up at the GitHub releases page, which is where you'll then read what _changed_ for a given version. The human-readable format is important here.

That said, I really like what JwageGenerateChangelog creates, as it provides _attribution_, which is also something I think it's important for us to do as a community OSS project.

On top of all that: I can see value in this tooling outside of Laminas. Yes, we're the primary consumers right now, but you've done a great job of engineering it such that it is generalized enough to work with any project that wants to adopt the workflow it implements for tagging, releasing, and up-merging. As such, we cannot necessarily assume that a given project has a changelog, much less in keep-a-changelog format.

So, here's my idea for moving forward:

- I'll split out the "add date to changelog and commit file" aspect into a separate utility that is passed to `ReleaseCommand` and invoked prior to retrieving the changelog text. This utility will be a no-op if no changelog file is present or if the changelog is not in keep-a-changelog format.

- I'll merge the `ReleaseChangelogAndFetchContents` into `CreateReleaseText` interfaces, adding arguments for source branch and repository directory to the latter, and make the current `UseKeepAChangelog....` implement that instead, having it return an empty string if no keep-a-changelog file is found. I'll also create an "appending aggregate" version that executes each and concatenates it to the previous.

I think this will resolve the architectural issues you've flagged.

Reply:
> KeepAChangelog does not generate a changelog; it is the changelog. The phly/keep-a-changelog tooling was being used to:

One is `f :: API -> Text`, the other is `f :: API -> (Text, GitCommit)` - very different underlying architecture

Reply:
> I think this will resolve the architectural issues you've flagged.

Will likely be able to do another review tonight :)

Reply:
Discussed in chat. It is clear now that this should be a 2 step operation of: 

1. generating the text (to be manipulated and/or used directly)
2. committing it to repo (separate service/function that does just that)